<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Retry</title>
</head>
<body>
    <h1>Retry</h1>

    <p>
        Write a function called withBoxUnlocked that takes a function value as argument, <br>
        unlocks the box, runs the function, and then ensures that the box is locked again <br>
        before returning, regardless of whether the argument function returned normally or <br>
        threw an exception.
    </p>
    <hr>
    <p>This exercise is logged into the developer tools console using console.log()</p>

    <script>
        var box = {
            locked: true,
            unlock: function () { this.locked = false; },
            lock  : function () { this.locked = true;  },
            _content: [],
            get content() {
                if (this.locked) throw new Error("Locked!");
                return this._content;
            }
        };

        function withBoxUnlocked (body) {
            var locked = box.locked;
            if (! locked) {
                return body();
            }

            box.unlock();

            try {
                return body();
            } finally {
                box.lock();
            }
        }

        withBoxUnlocked(function () {
            box.content.push("gold piece");
        });

        try {
            withBoxUnlocked(function () {
                throw new Error("Pirates on the horizon! Abort!");
            });
        } catch (e) {
            console.log("Error raised:", e);
        }

        console.log(box.locked);
    </script>
</body>
</html>